File: src/cmcandy/solutions/_0008.java
package cmcandy.solutions;


import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by CHEN on 2020/4/6.
 */
public class _0008 {


    static class Solution {
        public int myAtoi(String str) {
            char[] arr = str.toCharArray( );
            int len = arr.length;
            int index = 0;
            long ans = 0;
            int symbol = 0;
            boolean jud = true;
            while (index < len && arr[index] == ' ') {
                // 去掉前导空格
                index++;
            }
            while (jud && index < len) {

                if (arr[index] >= 48 && arr[index] <= 57) {
                    //if num
                    if (symbol == 0) {
                        symbol = 1;
                    }
                    ans = ans * 10 + (int) arr[index] - 48;
                } else if (arr[index] < 48 || arr[index] > 57) {
                    //if not num
                    if (ans == 0 && symbol == 0) {
                        if (arr[index] == '+') {
                            symbol = 1;
                        } else if (arr[index] == '-') {
                            symbol = -1;
                        } else {
                            jud = false;
                        }
                    } else {
                        jud = false;
                    }
                }
                if (ans > Integer.MAX_VALUE) {
                    //if >maxium
                    if (symbol == 1) {
                        ans = Integer.MAX_VALUE;

                    } else if (symbol == -1) {
                        ans = Integer.MIN_VALUE;
                        symbol = 1;
                    }
                    jud = false;
                }
                index++;
            }
            return (int) (symbol * ans);
        }

    }

    public static void main(String[] args) {
        Solution solution = new Solution( );
        System.out.println(solution.myAtoi("words and 987"));
    }
}

File: src/cmcandy/Python_language_Answers/_0008.py
class Solution:
    def myAtoi(self, s: str) -> int:
        length = len(s)
        index = 0
        symbol = 1
        num = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
        while index < length and s[index] == ' ':
            # 前导空格
            index += 1
        if index < length and s[index] == '+':
            # 正还是负号
            index += 1
        elif index < length and s[index] == '-':
            index += 1
            symbol = -1
        tmp = 0
        while index < length and s[index] in num:
            # 非数字字符或到达输入的结尾
            tmp *= 10
            tmp += int(s[index])
            index += 1
        # 如果整数数超过 32 位有符号整数范围，固定在边界值
        tmp *= symbol
        if tmp > 2**31 - 1:
            tmp = 2**31 - 1
        elif tmp < -2**31:
            tmp = -2**31
        return tmp


if __name__ == '__main__':
    s = Solution()
    print(s.myAtoi(' -42'))